rfc/0019-hashed-mojmap.md

```diff
@@ -0,0 +1,97 @@
+1. Class names are converted to `Cls_$x` where `$x` is the last 6 digits of the base-62 representation of the SHA-1 hash of the class name.
+
+The above steps are designed to preserve some of the benefits of intermediary as much as possible without the need for matching. Testing has shown that around 50% of class renames in Mojmap are only repackages, which are unlikely to affect hashed Mojmap. The descriptor removal is designed to preserve method names where only the descriptor has changed, while ensuring hashed Mojmap names are all unique.
+
+### Probability of collision
+
+As with any hash, a hash of Mojmap names is not guaranteed to be unique. Therefore, hash lengths were chosen such that the probability that we ever encounter such a collision within the lifetime of Quilt (and then some) falls below an acceptable level.
+
+The probability of any two hashes colliding is p = 1 - 2ⁿ! / (2ᵏⁿ (2ⁿ - k)!), where n is the number of bits in the hash, and k is the number of items we are hashing. Factorials of this size are hard to compute exactly, so we will instead work with the accurate approximation p = 1 - exp(-k²/2ⁿ⁺¹).
+
+To find out how many bits of the hash we need to get a certain probability, we can rearrange the formula: n = log₂(-k²/ln(1-p)) - 1.
+
+Using the estimate that there are 5000 classes in Minecraft, and that we want a probability of 0.001, we get n = 34.54 bits. Rounding up to 36, we find that we need 6 base-64 digits to achieve this probability, so 6 base-62 digits to achieve approximately this probability as well. Using an estimate of 40,000 fields/methods in Minecraft, and the same target probability, you can work out in a similar way that 7 digits is appropriate.
+
+If we got an entirely new set of mappings every snapshot, and snapshots were to occur on average once per week, then we would expect to get a class/field/method name collision about once every 192 years. But we don't get an entirely new set of mappings every snapshot, in reality the mappings are mostly the same from snapshot to snapshot, so it will take orders of magnitude longer than 192 years until we actually expect to run into any collision.
+
+### Compatibility with Fabric
+
+- Fabric mods will be remapped from intermediary to hashed Mojmap when they are loaded.
+- Quilt Loader's mapping resolver will recognize the `intermediary` channel when Fabric mods are loaded, and remap it to the current runtime mappings (`hashed_mojmap` in production). Quilt Loader will also recognize the `hashed_mojmap` channel.
+- The buildscript in Quilt's fork of Yarn will be updated to generate hashed Mojmap to Yarn mappings when built. When we start accepting PRs to our fork, the repository itself will be changed from using intermediary to hashed Mojmap. If necessary, a tool will be written to merge upstream from that point on.
+
+## Drawbacks
```

Comment:
We can work off of prior update experience with the `Block` -> `AbstractBlock` mass move of several methods. For yarn at least that would wreck havoc there based on the current hashing method also taking the owning class into account.

Do we want to try to mitigate that or leave yarn maintainers to sweeping the ash away.

Reply:
Agree with Cheater here. Rare cases like this which likely break binary compatibility anyway should be fixed in Yarn. That doesn't mean it has to be fixed manually though.